maze_dataset/tokenization/maze_tokenizer.py

```diff
@@ -388,3 +429,723 @@ def clear_cache(self):
+            legacy_maze_tokenizer = legacy_maze_tokenizer.tokenization_mode
+        return {
+            TokenizationMode.AOTP_UT_uniform: MazeTokenizer2(),
+            TokenizationMode.AOTP_UT_rasterized: MazeTokenizer2(),
+            TokenizationMode.AOTP_CTT_indexed: MazeTokenizer2(
+                coord_tokenizer=CoordTokenizers.CTT()
+            ),
+        }[legacy_maze_tokenizer]
+
+    def to_tokens(
+        self,
+        conn_list: ConnectionList,
+        origin: Coord | None,
+        target: Iterable[Coord],
+        path: CoordArray | None,
+    ) -> list[str]:
+        """Converts a set of maze elements into a list of tokens.
+        """
+        return self.prompt_sequencer.to_tokens(
+            conn_list,
+            origin,
+            target,
+            path,
+            self.coord_tokenizer,
+            self.adj_list_tokenizer,
+            self.target_tokenizer,
+            self.path_tokenizer,
+        )
+
+    @classmethod
+    def from_tokens(
+        cls,
+        tokens: str | list[str],
+    ) -> "MazeTokenizer2":
+        """
+        Infers most MazeTokenizer parameters from a full set of tokens.
+        Could be useful for adapting old code to new `MazeTokenizer`.
+        Would probably need a couple of other pieces of info besides just tokens.
+        - max_grid_size
+        - rasterization_mode: Only needed if UT tokens
+        - Anything else?
+        """
+        # Don't need directly, but something similar needed for LatticeMaze.from_tokens
+        raise NotImplementedError
+
+    @property
+    def token_arr(self) -> list[str] | None:
+        return VOCAB_LIST
+
+    @property
+    def tokenizer_map(self) -> dict[str, int]:
+        """map from token to index"""
+        return VOCAB_TOKEN_TO_INDEX
+
+    @property
+    def vocab_size(self) -> int:
+        return len(VOCAB_LIST)
+
+    @property
+    def n_tokens(self) -> int:
+        raise NameError(
+            "`MazeTokenizer2.n_tokens` has been removed. Use `len(maze_dataset.VOCAB_LIST)` instead."
+        )
+
+    @property
+    def padding_token_index(self) -> int:
+        return VOCAB_TOKEN_TO_INDEX[VOCAB.PADDING]
+
+    # conversion functions
+    # ============================================================
+
+    def coords_to_strings(self, coords: list[CoordTup | Coord]) -> list[str]:
+        return list(flatten([self.coord_tokenizer.to_tokens(c) for c in coords]))
+
+    @staticmethod
+    def strings_to_coords(
+        text: str,
+        when_noncoord: WhenMissing = "skip",
+    ) -> list[str | CoordTup]:
+        warnings.warn(
+            "`MazeTokenizer2.strings_to_coords` only supports legacy UT strings. Will be replaced by a more generalized function in a future release.",
+            PendingDeprecationWarning,
+        )
+        return strings_to_coords(text=text, when_noncoord=when_noncoord)
+
+    @staticmethod
+    def encode(text: str | list[str]) -> list[int]:
```

Comment:
glad we went with unified vocab here, makes encoding/decoding a whole lot easier!